### PR TITLE
Fix -Wall warnings

### PR DESCRIPTION
**Pull request by monoidal**
Refs #93 

This fixes most, but not all, warnings.

### CODE REVIEW

File: compiler/basicTypes/MkId.hs [hunk 1]
```diff
@@ -942,7 +942,7 @@ dataConArgUnpack
    ->  ( [(Weighted Type, StrictnessMark)]   -- Rep types
        , (Unboxer, Boxer) )
 
-dataConArgUnpack (Weighted arg_weight arg_ty)
+dataConArgUnpack (Weighted _ arg_ty)
```
Comment (aspiwack) [inline compiler/basicTypes/MkId.hs]: This one is suspicious. Is is correct?

File: compiler/basicTypes/Weight.hs [hunk 2]
```diff
@@ -121,12 +119,12 @@ instance Num (GMult a) where
...
-  ppr Omega = text "ω"
-  ppr (RigAdd m1 m2) = parens (ppr m1 <+> text "+" <+> ppr m2)
-  ppr (RigMul m1 m2) = parens (ppr m1 <+> text "*" <+> ppr m2)
-  ppr (RigThing t) = ppr t
+  ppr Zero_ = text "0"
+  ppr One_ = text "1"
+  ppr Omega_ = text "ω"
+  ppr (RigAdd_ m1 m2) = parens (ppr m1 <+> text "+" <+> ppr m2)
+  ppr (RigMul_ m1 m2) = parens (ppr m1 <+> text "*" <+> ppr m2)
+  ppr (RigThing_ t) = ppr t
```
Comment (aspiwack) [inline compiler/basicTypes/Weight.hs]: Why are the pattern synonyms wrong here? I'd rather use them: they're prettier.

File: compiler/types/Unify.hs [hunk 3]
```diff
@@ -1430,7 +1430,7 @@ ty_co_match menv subst ty1 (AppCo co2 arg2) _lkco _rkco
 
 ty_co_match menv subst (TyConApp tc1 tys) (TyConAppCo _ tc2 cos) _lkco _rkco
   = ty_co_match_tc menv subst tc1 tys tc2 cos
-ty_co_match menv subst (FunTy w ty1 ty2) co _lkco _rkco
+ty_co_match menv subst (FunTy _w ty1 ty2) co _lkco _rkco
```
Comment (aspiwack) [inline compiler/types/Unify.hs]: Mmm… this is wrong isn't it? We should be unifying the multiplicities, shouldn't we?

File: compiler/basicTypes/Weight.hs (hunk 2 above)
Comment (monoidal) [reply compiler/basicTypes/Weight.hs]: Fixed. Pattern synonyms were giving an incorrect incomplete pattern matching warning, and I wanted to silence it. I forgot that the COMPLETE pragma can be used for this.

Comment (aspiwack) [reply compiler/basicTypes/Weight.hs]: Ah. That's my bad. I somehow thought I'd done that already.

**Comment by aspiwack** [comment]
For both unresolved points: I restored the variable and used it in the logic. I'm fairly confident, let's see what the tests say.

**Pull request merged**